### PR TITLE
Initial upload business-tax-report.scm

### PR DESCRIPTION
**Pull request by christopherlam**
Report for periodic GST/VAT computation. I can accept bug reports.

### CODE REVIEW

**Comment by christopherlam** [comment]
Too difficult!

**Comment by gjanssens** [comment]
Huh, what's too difficult ?

**Comment by christopherlam** [comment]
Can't figure out why Travis keeps failing on some .po issue

**Comment by gjanssens** [comment]
The path you added to POTFILES.in doesn't match the actual location of the source file. The location of the report is wrong in any case. It should most likely be in gnucash/report/standard-reports (as you added in POTFILES.in).

What is this report doing exactly ? It's called "business-tax-report", but at first glance it's not using the business features (except for the Tax Tables, which are being used completely out of context). I'm not sure that's such a good idea...

It also seems to assume a specific way of entering transactions. Does it work on AR and AP accounts ?

**Comment by christopherlam** [comment]
It does work on AR & AP accounts using accrual accounting. I've tested today as per screenshots below, using relevant tax tables. The purpose of this report is summarise the GST/VAT generated periodically (quarterly in Australia) for reporting to ATO.

I'll welcome any ideas on how to select the relevant tax accounts.

business-tax-report

invoice

**Comment by gjanssens** [comment]
Ok, in that case I think your report input should be invoices and bills rather than transactions in the AR and AP accounts. You can study the different invoice reports to learn how the tax information is extracted from invoices/bills.

Considering it does use business data (and is called business-tax-report.scm) I would suggest the best location for your final report is gnucash/report/business-reports. The "business-" part of your report name can be omitted by the way. Note that loading business reports happens slightly differently from standard reports. To have it loaded from there, you should add a (use-modules (gnucash report tax-report.scm)) line to gnucash/report/business-reports/business-reports.scm (where all the other reports are loaded as well).

**Comment by christopherlam** [comment]
I personally do cash accounting only and diligently record all GST splits, and have hardly touched the business features at all in my years of using gnc, and this customised report is particularly valuable to me. I'll look into the Business invoices & bills. It's an interesting foray into scheme in any case.

I still think extracting transactional data is useful. Lots of transactions do not necessarily require the business features e.g. I'd record utility bills as regular transactions rather than monthly business bills and payments. But as you say perhaps this hybrid report does not really belong anywhere in the standard reports.

I'll keep you posted.

**Comment by gjanssens** [comment]
I'm trying to think with your here. From what I think I've understood so far you are combining aspects of gnucash that weren't designed to be combined.

So I'm trying to understand the concepts behind your report to figure out how to improve it. Since you have closed the PR and removed the path-1 branch from your own repo, you'll have to explain me :)
Where is your report is getting its data ? Which accounts are used to extract splits from ? Your sample report above only shows one line under an Income link. So I assume you were scanning the Income account to generate that report. Does the user have to select which accounts are used or did you hard-code it to income/expense accounts ?

**Comment by christopherlam** [comment]
Hi Geert

Scenario: My medical income is generated via billing clients & insurers. The income is unpredictable. The tax charged is also unpredictable, not a straightforward % because consultations are GST-free whereas reports attract GST. There are too many items (upwards of 200 per paycheck) to bother inputting into a Business Invoice.  Moreover the agency charges 40% (+10% GST) as a service fee on these billings.

A paycheck can show the following (assume GST 10%):
Income:Sales -$10000     (comprises $500 report and $9500 from clients and insurance companies)
GST Collected on sales = -$50 (because the report $500 attracted GST of 10%)
Expense:Agency $4000
GST Paid to agency = $400

According to Gnucash as strictly designed, I should generate an invoice, separate the GST-free billings of $9500 from the GST-charged billings $500, use appropriate tax table for each. I’ll need to process payment. I should also generate a bill to agency for $4000+$400, and immediately pay them.

But to me it’s much more straightforward to enter the data as a regular transaction with 4 splits, being careful to separate the numbers. Cash accounting works well. The agency is very reliable and will always pay me regularly (otherwise I walk!), therefore I don’t need to generate invoices/bills nor assign payments. Therefore I do not really require the Business features i.e. aging, jobs, process payments etc.

However I still need to track taxes being collected or paid – hence the use of tax tables as a placeholder to locate these tax accounts.

The report obtains data from INCOME and EXPENSE accounts, and will scan all transactions. Each transaction will be analysed, and if it contains a split to a GST account, it will be taken out and displayed into its own column. Hence sample report - and I'm particularly interested in the total GST sales and total GST on purchases in a particular period. I just find this format particularly intuitive because it links all relevant figures together rather than separate pages

Date......... Description........Tot.Sales | Net Sales | GST Sales | Tot.Purch | Net Purch  | GST Purch
03/09/17 AgencyName    $10050  | $10000    |         $50   |    $4400   |    $4000     |         $400
17/09/17 AgencyName    $11040  | $11000    |         $40   |    $4857  |     $4416     |         $441.60
Totals                                $21110  | $21000    |          $90   |    $9257  |     $8416     |         $841.60

GST receivable for this period = 841.60 - 90 = $751.60

**Comment by christopherlam** [comment]
Addendum the INCOME and EXPENSE accounts are not automatically picked, but are selected exactly like in transaction.scm

**Comment by gjanssens** [comment]
Ok that helps and I can imagine such a report to work for both accrual and cashed based accounting indeed.

First about your addendum: do you limit the list of selectable accounts to INCOME and EXPENSE accounts ? Or do you list all ? The latter would be problematic IMO because it may lead to wrong calculations of none income/expense accounts are selected. For example in the accrual accounting scenario if the user selects AR/AP in addition to INCOME/EXPENSE.

I think my main reservation with this report is in the hardwired requirement to generate to fake tax tables. That breaks in accrual accounting. In a way this would be a good thing, because your assumption to report only one total for each of income and expense is too restricted. I'd have to report 3-6 separate collections for each of income and expense. They are properly defined in my tax tables. Not hardcoding this does create a new challenge though: how will you know whether an account mentioned in a tax table entry is for expenses or income ? For simple transactions with only income/tax or expense/tax this can be deduced. But the multi-split transaction example you posted earlier has both and so you can't deduce this any more. With accrual accounting on the other hand you can because you have an invoice and a bill with their associated tax entries. It's more work but gives you the necessary information.

I'm not sure how to solve this right now. I think the tax tables should make a distinction between tax tables applicable to sales and purchases after all. (I'm comparing with other accounting systems I have access to for the Belgian jurisdiction and that's how they handle it).

**Comment by christopherlam** [comment]
Hi currently the whole account tree is displayed; with further tweaking I guess it could show INCOME/EXPENSE/AP/AR only.

W.R.T. deducing if a tax account relates to expense or income, currently it relies on whether it belongs to "Input" or "Output" tax tables. Perhaps it can instead rely on whether the tax account is of type Asset (relates to expense) or Liability (relates to income), as is currently being encouraged:
https://www.gnucash.org/docs/v2.6/C/gnucash-guide/appendixc_vat1.html

You may have noticed the report can handle multiple input/output tax accounts and can display them summed or individual-columns.

I'm trying to get my head around understanding 3-6 collections, hence my request for examples in mailinglist.

**Comment by christopherlam** [comment]
I've made further patches to (1) limit accounttype selection (2) move to standard-reports (3) toggle duplicate transaction removal.

I think I can solve the 'select taxtable' issue - will need to create a new selector (gnc:make-taxtable-list-option) in business-options.scm to select multiple taxtables, in the same style as (gnc:make-account-list-option) in options.scm. This will eliminate need for fake taxtables.

I'll then assign taxes to income/expense as per account type asset/liability as described above.

**Comment by christopherlam** [comment]
https://github.com/christopherlam/gnucash/blob/patch-1/gnucash/report/standard-reports/business-tax-report.scm

Some further changes were made
(1) removed requirement for fake tax tables by directly asking user to select tax accounts in options.
(2) I've forced the 'duplicate transaction removal'. This is required for correct calculations for transactions which contain multiple splits to the same account
(3) add option to display income/expense accounts in individual columns

I'm rather happy with these changes myself. I'll reopen - but perhaps need further tweaking prior to commit.

**Comment by christopherlam** [comment]
I think this is ready for beta testing. I've been trying to solve a problem - "cannot display $0.00 cells" which may be relevant to zero-rated GST jurisdictions. The cells and the collectors cannot distinguish between null cells and $0 cells, therefore, $0 amounts are simply disabled. However it would be nice if transactions which contain splits to a zero-rated GST account could be labelled $0.00, and subtotals should show $0.00 or empty-cell. I haven't been able to solve this.

**Comment by christopherlam** [comment]
I've managed to solve the $0.00-vs-null problem! This required rather annoying debugging. Looking at reports, I believe it would be preferable to limit accounts to INCOME and EXPENSE only instead of AP/AR/INCOME/EXPENSE. After all, acrual accounting means AP/AR will always involve transfer to INCOME/EXPENSE anyway. Moreover AP/AR will also include the subsequent payments to/from bank, which will simply render as dates without any numbers, because they don't involve income/expense/tax accounts. 

See the following sample reports for both filters (BTW - also notice the $0 cell from GBP Expenses on 07/09/17 being counted and subtotaled!) -- these report the exact same information, except the AP/AR accounts add additional clutter. 

Can merge onto patch-1. How does beta-testing work for this?

https://github.com/christopherlam/gnucash/blob/patch-1-wip-zero-cells/gnucash/report/standard-reports/business-tax-report.scm

## AP/AR/INC/EXP
image


## INC/EXP
image

**Comment by gjanssens** [comment]
> acrual accounting means AP/AR will always involve transfer to INCOME/EXPENSE anyway.

That's a wrong assumption. It is equally possible to enter bills and invoices that transfer between AP/AR and Assets/Liabilities. Imagine you receive a bill for something you have to charge through to a customer. You won't record this as an expense, but as an asset or a liability. When you write your customer's invoice you transfer again to the same asset/liability to neutralize it again. Still the bill and your invoice can contain taxes which you'd want to show up in your tax report.

> Moreover AP/AR will also include the subsequent payments to/from bank (see 6/9/17 entries in AP/AR sections)

Agreed. Payment transactions don't add to this report. You can filter them out because they are of a different transaction type than invoice/bill transactions. There can also be lot link transactions in the AR/AP accounts which are of no use to your report and hence can best be filtered out.

**Comment by christopherlam** [comment]
Would be grateful for opinion on this latest commit 'Overhaul Sorting tab' - it can be backported to transaction.scm without any compatibility issues. No new functionality, and should present sane sorting options to newcomers.

**Comment by christopherlam** [comment]
> Moreover AP/AR will also include the subsequent payments to/from bank (see 6/9/17 entries in AP/AR sections)
>> Agreed. Payment transactions don't add to this report. You can filter them out because they are of a different transaction type than invoice/bill transactions. There can also be lot link transactions in the AR/AP accounts which are of no use to your report and hence can best be filtered out.

Ok I have absolutely no idea how Lot Link txns work, so, if you have a sample datafile I'll be keen to see. Perhaps best leave it to display all transactions in AP/AR/INC/EXP. I wouldn't be keen to filter transactions *prior* to passing it through to (make-split-table) because I think it'd be unnecessarily expensive to run all calculations twice - first to detect useless splits, and second to pass through to (make-split-table).

**Comment by gjanssens** [comment]
A lot link transaction is a special kind of payment transaction. It is only 
used when a payment involves both a bill(invoice) and credit note.

Simple example:
A vendor sends you a wrong bill and afterwards you receive a credit note to 
compensate  for it. You book both documents in gnucash. At this point you have 
two unpaid documents (a bill and a credit note). You can have the credit note 
"pay" the bill by selecting both documents in the Bill payment window. Such a 
particular payment will create a lot link transaction.

If the bill and credit note were for the same amount, the lot link transaction 
will have two splits, both in the AP account.

If the bill was more than the credit note the procedure extends a little. In 
this case your payment in the Bill Payment window involves the two documents 
*and* an additional payment to some asset account (cash/bank/...).

In this case the lot link transaction will have 3 splits, of which two in the 
AP account (for the respective amounts of the bill and the credit note) and 
one to the asset account for the payment remainder).

**Comment by gjanssens** [comment]
As for how to handle those lot link transactions for your particular use case, just treat them exactly the same as you treat payment transactions in AR/AP. If you filter out payment transactions at some point, do the same for lot link transactions.

**Comment by christopherlam** [comment]
Ok that makes sense. I think it just means; from the splits list, remove
all splits which have no useful cells, ie tax, income or expense amounts
are all null. ($0 cells must be allowed though). Shouldn't be
computationally too onerous. I'll get on it.

**Comment by gjanssens** [comment]
A few git related suggestions as well:
- as with the other commit, please mind your commit messages. The first line should be some kind of subject line summarizing what the commit does. If you want to add more detail, do so starting on the third line (so leave one empty line between the "subject" and the "body" of the commit message). That's what most git related tools expect as commit message format.
- Can you rebase your work on current master (and do so frequently). This helps to avoid cluttering our master repositories by keeping branches to merge short (and hence easier to track).
- It would also be helpful to choose more relevant names for your feature branches. "patch-1" or "patch-2" don't hold much of a clue what feature the branch will carry. Note these names often end up in the final merge into our central repository so a branch name like "business-tax-report" or "txn-report-improvements" would be much more meaningful and useful. Thanks!

**Comment by christopherlam** [comment]
What a mess. Closing for now to fix this mess.

> When run on my test book, it spews lots of scheme errors, all of this form:
> 
> ;;; WARNING (wrong arguments for gnc:make-gnc-monetary: #<swig-pointer gnc_commodity * 41d9960> #< commodity: #<swig-pointer gnc_commodity * 41d9960> amount: #< num: 0 denom: 100>>)

I cannot reproduce this error. Perhaps you could send me a test book?

File: gnucash/report/standard-reports/business-tax-report.scm [hunk 1]
```diff
@@ -671,6 +671,31 @@ accounts must be of type ASSET for taxes paid on expenses, and type LIABILITY fo
...
+                  (N_ "Select regular transactions only."))
+          (vector (list TXN-TYPE-INVOICE)
+                  (N_ "Select invoices/bills only.")
+                  (N_ "Select invoices/bills only."))
+          (vector (list TXN-TYPE-PAYMENT)
+                  (N_ "Select payments only.")
+                  (N_ "Select payments only."))
+          (vector (list TXN-TYPE-LINK)
+                  (N_ "Select link transactions only.")
+                  (N_ "Select link transactions only.")))))
```
[the remark this answers is not in the document]
Comment (christopherlam) [reply gnucash/report/standard-reports/business-tax-report.scm]: Ok will delete, and also in the patched transaction.scm?

**Pull request closed without merging**